File: XamarinStore.Droid/Fragments/ProductDetailsFragment.cs
using System;
using System.Linq;

using Android.App;
using Android.OS;
using Android.Views;
using Android.Widget;
using Android.Content;
using System.Timers;
using System.Linq.Expressions;
using Android.Util;
using Android.Animation;
using Android.Graphics;

namespace XamarinStore
{
	class ProductDetailsFragment : Fragment, ViewTreeObserver.IOnGlobalLayoutListener
	{
		ImageView productImage;
		int currentIndex = 0;
		Product currentProduct;
		bool shouldAnimatePop;
		BadgeDrawable basketBadge;
		public Action<Product> AddToBasket = delegate {};
		string[] images = new string[0];
		bool cached;
		int slidingDelta;
		Spinner sizeSpinner;
		Spinner colorSpinner;

		KenBurnsDrawable productDrawable;
		ValueAnimator kenBurnsMovement;
		ValueAnimator kenBurnsAlpha;

		public ProductDetailsFragment ()
		{

		}

		public ProductDetailsFragment (Product product,int slidingDelta )
		{
			this.slidingDelta = slidingDelta;
			currentProduct = product;
			images = product.ImageUrls.ToArray().Shuffle() ?? new string[0];
		}

		public override void OnCreate (Bundle savedInstanceState)
		{
			base.OnCreate (savedInstanceState);
			RetainInstance = true;
			SetHasOptionsMenu (true);
		}

		public override View OnCreateView (LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState)
		{
			return inflater.Inflate (Resource.Layout.ProductDetail, null, true);
		}

		public override void OnViewCreated (View view, Bundle savedInstanceState)
		{
			base.OnViewCreated (view, savedInstanceState);

			productImage = View.FindViewById<ImageView> (Resource.Id.productImage);
			
			sizeSpinner = View.FindViewById<Spinner> (Resource.Id.productSize);

			colorSpinner = View.FindViewById<Spinner> (Resource.Id.productColor);

			var addToBasket = View.FindViewById<Button> (Resource.Id.addToBasket);
			addToBasket.Click += delegate {
				currentProduct.Size = currentProduct.Sizes [sizeSpinner.SelectedItemPosition];
				currentProduct.Color = currentProduct.Colors [colorSpinner.SelectedItemPosition];
				shouldAnimatePop = true;
				Activity.FragmentManager.PopBackStack();
				AddToBasket (currentProduct);
			};

			View.FindViewById<TextView> (Resource.Id.productTitle).Text = currentProduct.Name ?? string.Empty;
			View.FindViewById<TextView> (Resource.Id.productPrice).Text = currentProduct.PriceDescription ?? string.Empty;
			View.FindViewById<TextView> (Resource.Id.productDescription).Text = currentProduct.Description ?? string.Empty;

			((SlidingLayout)View).InitialMainViewDelta = slidingDelta;

			LoadOptions ();
		}

		void LoadOptions()
		{
			var sizeAdapter = new ArrayAdapter<ProductSize> (Activity, Android.Resource.Layout.SimpleSpinnerDropDownItem, currentProduct.Sizes);
			sizeAdapter.SetDropDownViewResource (Android.Resource.Layout.SimpleSpinnerDropDownItem);

			sizeSpinner.Adapter = sizeAdapter;
			sizeSpinner.SetSelection (currentProduct.Sizes.IndexOf (currentProduct.Size));

			var colorAdapter = new ArrayAdapter<ProductColor> (Activity, Android.Resource.Layout.SimpleSpinnerDropDownItem, currentProduct.Colors);
			colorAdapter.SetDropDownViewResource (Android.Resource.Layout.SimpleSpinnerDropDownItem);

			colorSpinner.Adapter = colorAdapter;
		}

		public override void OnStart ()
		{
			base.OnStart ();
			AnimateImages ();
		}

		public override void OnStop ()
		{
			base.OnStop ();
			if (kenBurnsAlpha != null)
				kenBurnsAlpha.Cancel ();
			if (kenBurnsMovement != null)
				kenBurnsMovement.Cancel ();
		}

		public override void OnCreateOptionsMenu (IMenu menu, MenuInflater inflater)
		{
			inflater.Inflate (Resource.Menu.menu, menu);
			var cartItem = menu.FindItem (Resource.Id.cart_menu_item);
			cartItem.SetIcon ((basketBadge = new BadgeDrawable (cartItem.Icon)));

			var order = WebService.Shared.CurrentOrder;
			basketBadge.Count = order.Products.Count;
			order.ProductsChanged += (sender, e) => {
				basketBadge.SetCountAnimated (order.Products.Count);
			};
			base.OnCreateOptionsMenu (menu, inflater);
		}

		public override Android.Animation.Animator OnCreateAnimator (FragmentTransit transit, bool enter, int nextAnim)
		{
			if (!enter && shouldAnimatePop)
				return AnimatorInflater.LoadAnimator (View.Context, Resource.Animation.add_to_basket_in);
			return base.OnCreateAnimator (transit, enter, nextAnim);
		}

		void AnimateImages ()
		{
			if (images.Length < 1)
				return;
			if (images.Length == 1) {
				//No need to await the change
				#pragma warning disable 4014
				Images.SetImageFromUrlAsync (productImage, Product.ImageForSize (images [0], Images.ScreenWidth));
				#pragma warning restore 4014
				return;
			}

			productImage.ViewTreeObserver.AddOnGlobalLayoutListener (this);
		}

		public async void OnGlobalLayout ()
		{
			productImage.ViewTreeObserver.RemoveGlobalOnLayoutListener (this);

			const int DeltaX = 100;

			var img1 = Images.FromUrl (Product.ImageForSize (images [0], Images.ScreenWidth));
			var img2 = Images.FromUrl (Product.ImageForSize (images [1], Images.ScreenWidth));

			productDrawable = new KenBurnsDrawable (Color.DarkBlue);
			productDrawable.FirstBitmap = await img1;
			productDrawable.SecondBitmap = await img2;
			productImage.SetImageDrawable (productDrawable);
			currentIndex++;

			var evaluator = new MatrixEvaluator ();
			var finalMatrix = new Matrix ();
			finalMatrix.SetTranslate (-DeltaX, -(float)productDrawable.FirstBitmap.Height / 1.3f + (float)productImage.Height);
			finalMatrix.PostScale (1.27f, 1.27f);
			kenBurnsMovement = ValueAnimator.OfObject (evaluator, new Matrix (), finalMatrix);
			kenBurnsMovement.Update += (sender, e) => productDrawable.SetMatrix ((Matrix)e.Animation.AnimatedValue);
			kenBurnsMovement.SetDuration (14000);
			kenBurnsMovement.RepeatMode = ValueAnimatorRepeatMode.Reverse;
			kenBurnsMovement.RepeatCount = ValueAnimator.Infinite;
			kenBurnsMovement.Start ();

			kenBurnsAlpha = ObjectAnimator.OfInt (productDrawable, "alpha", 0, 0, 0, 255, 255, 255);
			kenBurnsAlpha.SetDuration (kenBurnsMovement.Duration);
			kenBurnsAlpha.RepeatMode = ValueAnimatorRepeatMode.Reverse;
			kenBurnsAlpha.RepeatCount = ValueAnimator.Infinite;
			kenBurnsAlpha.AnimationRepeat += (sender, e) => NextImage ();
			kenBurnsAlpha.Start ();
		}

		async void NextImage ()
		{
			currentIndex = (currentIndex + 1) % images.Length;
			var image = images [currentIndex];
			await Images.SetImageFromUrlAsync (productDrawable, Product.ImageForSize (image, Images.ScreenWidth));
			PrecacheNextImage ();
		}

		void PrecacheNextImage()
		{
			if (currentIndex + 1 >= images.Length)
				cached = true;
			if (cached)
				return;
			var next = currentIndex + 1;
			var image = images [next];
			//No need to await the precache to finish
			#pragma warning disable 4014
			FileCache.Download (Product.ImageForSize (image, Images.ScreenWidth));
			#pragma warning restore 4014
		}
	}
}

File: F#/XamarinStore.Droid/Fragments/ProductDetailsFragment.fs
namespace XamarinStore

open System
open System.Linq

open Android.App
open Android.OS
open Android.Views
open Android.Widget
open Android.Animation
open Android.Graphics
open Product
open Helpers
open Order
open System.Threading.Tasks

type ProductDetailsFragment(product:Product, slidingDelta) as this =
    inherit Fragment()

    let productImage : ImageView = null
    let mutable currentProduct = ref product
    let images = product.GetImageUrls().ToArray().RandomArrayPermutation()

    let productDrawable = new KenBurnsDrawable(Color.DarkBlue.ToAndroidColor())
    let mutable kenBurnsMovement: ValueAnimator = null
    let mutable kenBurnsAlpha:ValueAnimator = null

    let mutable currentIndex = 0
    let mutable productImage:ImageView = null
    let mutable shouldAnimatePop = false
    let mutable basketBadge: BadgeDrawable option = None
    let mutable sizeSpinner:Spinner = null
    let mutable colorSpinner:Spinner = null

    let mutable cached = false

    let PrecacheNextImage () =
        if (currentIndex + 1 >= images.Length) then
            cached <- true
        else if not cached then
            let next = currentIndex + 1
            let image = images.[next]
            //No need to await the precache to finish
            FileCache.Download (Product.ImageForSize (image, Images.ScreenWidth)) |> Async.Ignore |> Async.StartImmediate

    let NextImage () = async {
        currentIndex <- (currentIndex + 1) % images.Length
        let image = images.[currentIndex]
        do! Images.SetBitmapImageFromUrlAsync productDrawable (Product.ImageForSize(image, Images.ScreenWidth))
        PrecacheNextImage () } |> Async.StartImmediate

    let AnimateImages () =
        if images.Length = 1 then
            Images.SetImageViewFromUrlAsync productImage (Product.ImageForSize(images.[0], Images.ScreenWidth))
            |> Async.StartImmediate
        if images.Length > 1 then
            productImage.ViewTreeObserver.AddOnGlobalLayoutListener (this)

    interface ViewTreeObserver.IOnGlobalLayoutListener with
        member this.OnGlobalLayout() = 
            async{ productImage.ViewTreeObserver.RemoveGlobalOnLayoutListener this
                   let! img1 = Images.FromUrl(Product.ImageForSize (images.[0], Images.ScreenWidth))

                   let! img2 = Images.FromUrl (Product.ImageForSize (images.[1], Images.ScreenWidth))

                   productDrawable.FirstBitmap <- img1
                   productDrawable.SecondBitmap <- img2

                   productImage.SetImageDrawable productDrawable

                   currentIndex <- currentIndex + 1
                      
                   let evaluator = new MatrixEvaluator ()
                   let finalMatrix = new Matrix ()

                   finalMatrix.SetTranslate (-100.0f, -(float32 productDrawable.FirstBitmap.Height) / 1.3f + (float32 productImage.Height))

                   finalMatrix.PostScale (1.27f, 1.27f) |> ignore
                   kenBurnsMovement <- ValueAnimator.OfObject (evaluator, new Matrix (), finalMatrix)
                   kenBurnsMovement.Update.Add(fun x-> productDrawable.SetMatrix (x.Animation.AnimatedValue :?> Matrix))
                   kenBurnsMovement.SetDuration 14000L |> ignore
                   kenBurnsMovement.RepeatMode <- ValueAnimatorRepeatMode.Reverse
                   kenBurnsMovement.RepeatCount <- ValueAnimator.Infinite
                   kenBurnsMovement.Start()
                     
                   kenBurnsAlpha <- ObjectAnimator.OfInt (productDrawable, "alpha", 0, 0, 0, 255, 255, 255)
                   kenBurnsAlpha.SetDuration kenBurnsMovement.Duration |> ignore
                   kenBurnsAlpha.RepeatMode <- ValueAnimatorRepeatMode.Reverse
                   kenBurnsAlpha.RepeatCount <- ValueAnimator.Infinite
                   kenBurnsAlpha.AnimationRepeat.Add(fun x-> NextImage())
                   kenBurnsAlpha.Start()
                   }

            |> Async.StartImmediate
            |> ignore

    member val AddToBasket = ignore with get,set

    override this.OnCreate savedInstanceState =
        base.OnCreate savedInstanceState
        this.RetainInstance <- true
        this.SetHasOptionsMenu true

    override this.OnCreateView(inflater, container, savedInstanceState) =
        inflater.Inflate(Resource_Layout.ProductDetail, null, true)


    member this.LoadOptions() =
        let sizeAdapter = new ArrayAdapter<ProductSize>(this.Activity, Android.Resource.Layout.SimpleSpinnerDropDownItem, (!currentProduct).Sizes.ToArray())
        sizeAdapter.SetDropDownViewResource (Android.Resource.Layout.SimpleSpinnerDropDownItem)

        sizeSpinner.Adapter <- sizeAdapter
        sizeSpinner.SetSelection ((!currentProduct).Sizes |> Array.findIndex(fun x->x = ((!currentProduct).Size)))

        let colorAdapter = new ArrayAdapter<ProductColor>(this.Activity, Android.Resource.Layout.SimpleSpinnerDropDownItem, (!currentProduct).Colors.ToArray())
        colorAdapter.SetDropDownViewResource (Android.Resource.Layout.SimpleSpinnerDropDownItem)

        colorSpinner.Adapter <- colorAdapter

    override this.OnViewCreated(view, savedInstanceState) =
        base.OnViewCreated(view, savedInstanceState)

        productImage <- this.View.FindViewById<ImageView> Resource_Id.productImage
        sizeSpinner <- this.View.FindViewById<Spinner> Resource_Id.productSize
        colorSpinner <- this.View.FindViewById<Spinner> Resource_Id.productColor

        let addToBasket = this.View.FindViewById<Button> Resource_Id.addToBasket
        addToBasket.Click.Add(fun x->
            currentProduct := { (!currentProduct) with 
                                    Size = (!currentProduct).Sizes.[sizeSpinner.SelectedItemPosition]
                                    Color = (!currentProduct).Colors.[colorSpinner.SelectedItemPosition] }
            shouldAnimatePop <- true
            this.Activity.FragmentManager.PopBackStack()
            this.AddToBasket !currentProduct)

        this.View.FindViewById<TextView>(Resource_Id.productTitle).Text <- (!currentProduct).Name
        this.View.FindViewById<TextView>(Resource_Id.productPrice).Text <- (!currentProduct).GetPriceDescription()
        this.View.FindViewById<TextView>(Resource_Id.productDescription).Text <- (!currentProduct).Description

        (this.View :?> SlidingLayout).InitialMainViewDelta <- slidingDelta

        this.LoadOptions ()

    override this.OnStart () =
        base.OnStart ()
        AnimateImages ()

    override this.OnStop () =
        base.OnStop ()
        if not (kenBurnsAlpha = null) then kenBurnsAlpha.Cancel ()
        if not (kenBurnsMovement = null) then kenBurnsMovement.Cancel ()

    override this.OnCreateOptionsMenu (menu, inflater) =
        inflater.Inflate (Resource_Menu.menu, menu)
        let cartItem = menu.FindItem (Resource_Id.cart_menu_item)
        let tempBasketBadge = new BadgeDrawable (cartItem.Icon)
        basketBadge <- Some tempBasketBadge
        cartItem.SetIcon tempBasketBadge |> ignore

        let order = WebService.CurrentOrder
        tempBasketBadge.Count <- order.Products.Length
        WebService.CurrentOrder <- { order with Notification = Some (fun (x:Order) -> tempBasketBadge.SetCountAnimated (x.Products.Length)) }
        base.OnCreateOptionsMenu (menu, inflater)

    override this.OnCreateAnimator(transit, enter, nextAnim) =
        if not enter && shouldAnimatePop then
            AnimatorInflater.LoadAnimator (this.View.Context, Resource_Animation.add_to_basket_in) |> ignore
        base.OnCreateAnimator (transit, enter, nextAnim)
